File: src/main/java/self/ed/util/SudokuUtils.java
package self.ed.util;

import static java.lang.Integer.parseInt;
import static java.util.Arrays.stream;
import static java.util.stream.Collectors.joining;

import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;

public class SudokuUtils {

  private static final String EMPTY_AS_STRING = ".";

  public static String toString1D(int[][] matrix) {
    return stream(matrix)
        .map(line -> stream(line)
            .mapToObj(SudokuUtils::cellToString)
            .collect(joining()))
        .collect(joining());
  }

  public static String toString2D(int[][] matrix) {
    return stream(matrix)
        .map(line -> stream(line)
            .mapToObj(SudokuUtils::cellToString)
            .collect(joining(" ")))
        .collect(joining("\n"));
  }

  public static int[][] fromString1D(String flat) {
    String[] values = flat.split("");
    int size = (int) Math.sqrt(values.length);
    int[][] matrix = new int[size][size];
    for (int row = 0; row < size; row++) {
      for (int col = 0; col < size; col++) {
        String value = values[row * size + col];
        matrix[row][col] = cellFromString(value);
      }
    }
    return matrix;
  }

  public static int[][] fromString2D(String file) {
    return stream(file.split("\n"))
        .map(String::trim)
        .map(SudokuUtils::parseLine2D)
        .toArray(int[][]::new);
  }

  public static long countOpen(int[][] matrix) {
    return stream(matrix).mapToLong(line -> stream(line).filter(v -> v != 0).count()).sum();
  }

  public static int[][] copy(int[][] source) {
    int[][] target = new int[source.length][];
    for (int row = 0; row < source.length; row++) {
      target[row] = Arrays.copyOf(source[row], source[row].length);
    }
    return target;
  }

  private static int[] parseLine2D(String line) {
    return stream(line.split(" "))
        .map(String::trim)
        .mapToInt(SudokuUtils::cellFromString)
        .toArray();
  }

  private static int cellFromString(String cell) {
    return EMPTY_AS_STRING.equals(cell) ? 0 : parseInt(cell);
  }

  private static String cellToString(int cell) {
    return cell == 0 ? EMPTY_AS_STRING : String.valueOf(cell);
  }

  public static String getCurrentTime() {
    return new SimpleDateFormat("yyyyMMdd-HHmmss").format(new Date());
  }

  public static String join(String delimiter, Object... items) {
    return stream(items).map(Object::toString).collect(joining(delimiter));
  }
}

File: cpp-sudoku/src/util/SudokuUtils.cpp
#include "SudokuUtils.hpp"
#include <cmath>
#include <iostream>

namespace just::demo::util {

std::string SudokuUtils::toString1D(const std::vector<std::vector<int>>& matrix) {
    std::string result;
    for (const auto& row : matrix) {
        for (int cell : row) {
            result += cellToString(cell);
        }
    }
    return result;
}

std::string SudokuUtils::toString2D(const std::vector<std::vector<int>>& matrix) {
    std::string result;
    for (size_t i = 0; i < matrix.size(); ++i) {
        for (size_t j = 0; j < matrix[i].size(); ++j) {
            result += cellToString(matrix[i][j]);
            if (j < matrix[i].size() - 1) {
                result += " ";
            }
        }
        if (i < matrix.size() - 1) {
            result += "\n";
        }
    }
    return result;
}

std::vector<std::vector<int>> SudokuUtils::fromString1D(const std::string& flat) {
    std::vector<std::string> values;
    for (char c : flat) {
        values.emplace_back(1, c);
    }
    
    int size = static_cast<int>(std::sqrt(values.size()));
    std::vector<std::vector<int>> matrix(size, std::vector<int>(size));
    
    for (int row = 0; row < size; row++) {
        for (int col = 0; col < size; col++) {
            std::string value = values[row * size + col];
            matrix[row][col] = cellFromString(value);
        }
    }
    return matrix;
}

std::vector<std::vector<int>> SudokuUtils::fromString2D(const std::string& file) {
    std::istringstream iss(file);
    std::string line;
    std::vector<std::vector<int>> matrix;
    
    while (std::getline(iss, line)) {
        if (!line.empty()) {
            matrix.push_back(parseLine2D(line));
        }
    }
    return matrix;
}

long SudokuUtils::countOpen(const std::vector<std::vector<int>>& matrix) {
    long count = 0;
    for (const auto& row : matrix) {
        count += std::count_if(row.begin(), row.end(), [](int v) { return v != 0; });
    }
    return count;
}

std::vector<std::vector<int>> SudokuUtils::copy(const std::vector<std::vector<int>>& source) {
    std::vector<std::vector<int>> result;
    result.reserve(source.size());
    for (const auto& row : source) {
        result.push_back(row);
    }
    return result;
}

std::vector<int> SudokuUtils::parseLine2D(const std::string& line) {
    std::istringstream iss(line);
    std::string token;
    std::vector<int> result;
    
    while (iss >> token) {
        result.push_back(cellFromString(token));
    }
    return result;
}

int SudokuUtils::cellFromString(const std::string& cell) {
    return (cell == EMPTY_AS_STRING) ? 0 : std::stoi(cell);
}

std::string SudokuUtils::cellToString(int cell) {
    return (cell == 0) ? EMPTY_AS_STRING : std::to_string(cell);
}

} // namespace just::demo::util
